cou/apps/auxiliary.py

```diff
@@ -457,3 +462,146 @@ async def _verify_nova_compute(self, target: OpenStackRelease) -> None:
             raise ApplicationError(
                 f"Units '{', '.join(units_not_upgraded)}' did not reach {target}."
             )
+
+
+@AppFactory.register_application(["vault"])
+class Vault(AuxiliaryApplication):
+    """Application for vault."""
+
+    wait_timeout = LONG_IDLE_TIMEOUT
+    wait_for_model = True
+
+    def _get_cacert_file(self) -> Optional[str]:
+        """Read cert file and write into temporary file.
+
+        :return: Temporary file path
+        :rtype: str
+        """
+        cacert_b64 = self.config["ssl-ca"].get("vault")
+        cacert_file = None
+        if cacert_b64:
+            with tempfile.NamedTemporaryFile(mode="wb", delete=False) as fp:
+                fp.write(base64.b64decode(cacert_b64))
+                cacert_file = fp.name
+        return cacert_file
+
+    async def _get_unit_api_url(self, unit_name: str) -> str:
+        """Get unit's vault api address.
+
+        :param unit_name: vault unit name
+        :type unit_name: str
+        :return: unit's vault api url
+        :rtype: str
+        """
+        # Use vip as address if vip is being used.
+        if self.config["vip"].get("value"):
+            address = self.config["vip"].get("value")
+        else:
+            juju_unit = await self.model.get_unit(unit_name)
+            address = juju_unit.public_address
+
+        transport = "http"
+        if self.config["ssl-cert"].get("value"):
+            transport = "https"
+        return f"{transport}://{address}:8200"
+
+    async def _wait_for_sealed_status(self) -> None:
+        """Wait for application vault go into sealed status.
+
+        :raises ApplicationError: When application vault is not in sealed status.
+        """
+        await self.model.wait_for_idle(
+            timeout=self.wait_timeout,
+            status="blocked",
+            apps=[self.name],
+            # Vault application will go into error status first then blocked status.
+            raise_on_error=False,
+        )
+
+        app_status = await self.model.get_application_status(charm_name="vault")
+        if not app_status.status.info == "Unit is sealed":
+            raise ApplicationError("Vault not in sealed status")
```

Comment:
Can we update the behavior of `check_vault_status`, so we can use it here? WDYT?

Reply:
The reason I don't use `check_vault_status` here is because it belong to `steps.vault` module. I can move it to the Vault class and make it become the *classmethod*. Then I have to add error handling here. I feel it's nice to have, not required.

Reply:
Like this you created helper function which is used once and here you have lines which are similar to that function. Changing that function to `is_vault_sealed` and you can use it in both places.

Reply:
It still a little bit different. `check_vault_status` raise if it's in sealed status, `_wait_for_sealed_status` raise if it's not sealed status. The lines we save only two or three.

Reply:
I know, that's why I'm saying if you improved it, you can use it in both places.
e.g.
```pyhon
# here
if await is_vault_sealed(app) is False:
    raise ApplicationError("Vault not in sealed status")
    
# check_vault_status
if await is_vault_sealed(app) is True:
    raise VaultSealed(
        "Vault is in sealed, please follow the steps on "
        "https://charmhub.io/vault to unseal the vault manually before upgrade"
    )
```

Two more comments:
 - Here you are raising ApplicationError, but in check_vault_status you are using VaultSealed
 - Why passing charm name to get_application_status `app_status = await self.model.get_application_status(charm_name="vault")` instead getting status for this app? There should not be two vault apps, but if they are, we should raise an exception for proper one.

Reply:
> Here you are raising ApplicationError, but in check_vault_status you are using VaultSealed

Yes, or they should both be an `ApplicationError`. `VaultSealed` is more precise and only raise on this pre-check. Or I misunderstand what you want me to change?

> Why passing charm name to get_application_status app_status = await self.model.get_application_status(charm_name="vault") instead getting status for this app? There should not be two vault apps, but if they are, we should raise an exception for proper one.

I see your point, I can change this part.

My point is a `@classmethod` belong to `Vault` o7k application class should not be used in `steps` package. I feel we miss a module that provides helper function for specific juju application, which not belong to either `apps` or `steps` package.

Sorry this make me hesitate because it break my understand about the module levels in this project. And we don't get too much benefit from this practice. I suggest we keep what we have right now.

Reply:
> Yes, or they should both be an ApplicationError. VaultSealed is more precise and only raise on this pre-check. Or I misunderstand what you want me to change?

Both should be `VaultSealed`? Here I'm asking, for reason why one is `ApplicationError` and second `VaultSealed`.

> Please check function check_vault_status, in this case you won't know the application name.

We are in the Vault application, so you can do `self.charm` to get charm name instead of using `"vault"` string, however I think it would be more propriety to get status for application only, so using `self.name`.

> My point is a @classmethod belong to Vault o7k application class should not be used in steps package. I feel we miss a module that provides helper function for specific juju application, which not belong to either apps or steps package.
Sorry this make me hesitate because it break my understand about the module levels in this project. And we don't get too much benefit from this practice. I suggest we keep what we have right now.

You can put `is_vault_sealed` in to utils and also `check_vault_status`, or `check_vault_status` could be in steps/vault.py as you did.  The module structure is not good, I know.

Reply:
> Both should be VaultSealed? Here I'm asking, for reason why one is ApplicationError and second VaultSealed.

The VaultSealed raise because we don't support to upgrade cloud which vault in sealed status. Second case is that we want to reach the sealed status. It's wired we raise same error on the opposite situation.

> We are in the Vault application, so you can do self.charm to get charm name instead of using "vault" string, however I think it would be more propriety to get status for application only, so using self.name.

The `check_vault_status` in `steps.vault` module, it's not in the application.

Reply:
> The VaultSealed raise because we don't support to upgrade cloud which vault in sealed status. Second case is that we want to reach the sealed status. It's weird we raise same error on the opposite situation.

Ok, make sense.

> The check_vault_status in steps.vault module, it's not in the application.

I'm speaking about this line:
`app_status = await self.model.get_application_status(charm_name="vault")`

This will get any vault, what I'm suggesting is to use only current vault app status.

Reply:
> This will get any vault, what I'm suggesting is to use only current vault app status.

Any vault should be fine, I don't think we will have a case that will have multiple vault applications.